xla/ffi/api/ffi.h

```diff
@@ -110,7 +110,7 @@ namespace internal {
 
 template <DataType dtype>
 struct PtrType {
-  static_assert(sizeof(dtype) == 0, "unsupported data type");
+  static_assert(sizeof(dtype) == 1, "unsupported data type");
```

Comment:
This static assert is here to check that this template is never instantiated

Reply:
Well that explains why clang never complains about it either way. Unfortunately gcc fails it.

'''
ERROR: /home/andrew/src/tf_test/openxla/xla/ffi/api/BUILD:67:12: Compiling xla/ffi/api/ffi_test.cc failed: (Exit 1): gcc failed: error executing command (from target //xla/ffi/api:ffi_test) /usr/local/bin/gcc -U_FORTIFY_SOURCE -fstack-protector -Wall -Wunused-but-set-parameter -Wno-free-nonheap-object -fno-omit-frame-pointer -g0 -O2 '-D_FORTIFY_SOURCE=1' -DNDEBUG -ffunction-sections ... (remaining 203 arguments skipped)

Use --sandbox_debug to see verbose messages from the sandbox and retain the sandbox build root for debugging
In file included from xla/ffi/api/ffi_test.cc:16:
./xla/ffi/api/ffi.h:113:31: error: static assertion failed: unsupported data type
  113 |   static_assert(sizeof(dtype) == 0, "unsupported data type");
      |                 ~~~~~~~~~~~~~~^~~~
./xla/ffi/api/ffi.h:136:3: error: need 'typename' before 'xla::ffi::internal::PtrType<dtype>::Type' because 'xla::ffi::internal::PtrType<dtype>' is a dependent scope
  136 |   internal::PtrType<dtype>::Type* data;
      |   ^~~~~~~~
      |   typename 
./xla/ffi/api/ffi.h: In static member function 'static std::optional<xla::ffi::BufferBase<dtype> > xla::ffi::ArgDecoding<xla::ffi::BufferBase<dtype> >::Decode(XLA_FFI_ArgType, void*)':
./xla/ffi/api/ffi.h:152:30: error: need 'typename' before 'xla::ffi::internal::PtrType<dtype>::Type' because 'xla::ffi::internal::PtrType<dtype>' is a dependent scope
  152 |     auto* data = static_cast<internal::PtrType<dtype>::Type*>(buf->data);
      |                              ^~~~~~~~
      |                              typename 
./xla/ffi/api/ffi.h:152:60: error: expected '>' before '*' token
  152 |     auto* data = static_cast<internal::PtrType<dtype>::Type*>(buf->data);
      |                                                            ^
./xla/ffi/api/ffi.h:152:60: error: expected '(' before '*' token
  152 |     auto* data = static_cast<internal::PtrType<dtype>::Type*>(buf->data);
      |                                                            ^
      |                                                            (
./xla/ffi/api/ffi.h:152:61: error: expected primary-expression before '>' token
  152 |     auto* data = static_cast<internal::PtrType<dtype>::Type*>(buf->data);
      |                                                             ^
./xla/ffi/api/ffi.h:152:73: error: expected ')' before ';' token
  152 |     auto* data = static_cast<internal::PtrType<dtype>::Type*>(buf->data);
      |                                                                         ^
      |                                                                         )
In file included from external/com_google_googletest/googletest/include/gtest/gtest.h:67,
                 from external/tsl/tsl/platform/test.h:23,
                 from external/tsl/tsl/lib/core/status_test_util.h:20,
                 from xla/ffi/api/ffi_test.cc:27:
xla/ffi/api/ffi_test.cc: In lambda function:
xla/ffi/api/ffi_test.cc:66:22: error: 'struct xla::ffi::BufferBase<xla::ffi::DataType::F32>' has no member named 'data'
   66 |     EXPECT_EQ(buffer.data, storage.data());
      |                      ^~~~
INFO: Elapsed time: 979.624s, Critical Path: 735.21s
INFO: 6717 processes: 1840 internal, 4877 linux-sandbox.
FAILED: Build did NOT complete successfully
'''

Reply:
`sizeof(dtype) == 0` is a work around from https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2022/p2593r0.html, but looks like it might not work for non-type template parameters, can you try to remove default template definition

Reply:
Thanks for that link. I adopted the last workaround suggested that should be OK.